lib/parameter_parser/search_parameter_parser.rb

```diff
@@ -145,49 +145,48 @@ def return_fields
     fields
   end
 
-  def parameters_starting_with(prefix)
-    with_prefix = @params.select do |name, _|
-      name.start_with?(prefix)
-    end
-
-    with_prefix.each_with_object({}) do |(name, value), result|
-      @used_params << name
-      result[name.sub(prefix, "")] = value
-    end
-  end
-
-  def validate_filter_parameters(parameters, type)
-    allowed, disallowed = parameters.partition { |field, _|
-      allowed_filter_fields.include?(field)
+  def validate_filter_parameters(parameter_hashes)
+    allowed, disallowed = parameter_hashes.partition { |parameter_hash|
+      allowed_filter_fields.include?(parameter_hash['name'])
     }
 
-    disallowed.each do |field, _|
-      @errors << %{"#{field}" is not a valid #{type} field}
+    disallowed.each do |parameter_hash|
+      @errors << %{"#{parameter_hash['name']}" is not a valid #{parameter_hash['operation']} field}
     end
 
     allowed
   end
 
-  def filters
-    filter_parameters = parameters_starting_with("filter_")
-    reject_parameters = parameters_starting_with("reject_")
-
-    build_filters(
-      validate_filter_parameters(filter_parameters, "filter"),
-      validate_filter_parameters(reject_parameters, "reject")
-    )
+  # Returns hashes of
+  #   {'operation'=> 'filter/reject',
+  #    'name'=>'...',
+  #    'full_name'=>'...',
+  #    'values'=>[...]}
+  def analyze_filters(parameters)
+    matches = parameters.keys.map do |name|
+      /^(?'full_name'(?'operation'filter|reject)_(?'name'.*))$/.match(name)
+    end
+    matches.compact.map do |match|
+      captures = match.named_captures
+      full_name = captures['full_name']
+      captures.merge({ 'values' => parameters[full_name] })
+    end
```

Comment:
True, but now it is clearer this method depends on parameters explicitly - it has no other dependencies. Otherwise it would be a parameterless method and in my opinion it would be less clear what it does.

Reply:
OK, that's fair enough :smiley: